### PR TITLE
perf: use ets table for client ID registration

### PR DESCRIPTION
**Pull request by zmstone**
This is a performance optimization incorporated with a refactoring.

## Performance fix
`supervisor:which_children(wolff_client_sup)` becomes quite expensive when there are many clients started (thousands).
Main caller is `wolff:check_conectivity/1`.

The fix is to store client ID -> Pid in a ets table.

## Refactoring

1.9.0 made use of the producer name as ETS table key for producer process registration.
However that lead to two minor problems:
1. Users are still forced to name producer (although no longer has to be an atom)
2. Name clash is not checked, it might be the case that more than one `{ClientId, Topic}` uses the same producer name.
  the fix is to use `{ClientId, Topic}` as the ID, and the old "name" is only used as process registry (when it's an atom) for backward compatibility.

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+* 1.9.1
+  - Use ETS (named `wolff_clients_global`) for client ID registration.
+    When there are thousands of clients, supervisor:which_children becomse quite expensive.
```
Comment (thalesmg) [inline changelog.md:3]:
```suggestion
    When there are thousands of clients, `supervisor:which_children` becomes quite expensive.
```

File: src/wolff.erl [hunk 2]
```diff
@@ -27,6 +27,8 @@
 %% Supervised producer management APIs
 -export([ensure_supervised_producers/3,
          stop_and_delete_supervised_producers/1,
+         stop_and_delete_supervised_producers/2,
+         %% /2 is deprecated, call /2 instead
```
Comment (thalesmg) [inline src/wolff.erl:31]:
```suggestion
         %% /3 is deprecated, call /2 instead
```
?

File: src/wolff_client.erl [hunk 3]
```diff
@@ -146,7 +146,8 @@ recv_leader_connection(Client, Topic, Partition, Pid) ->
 delete_producers_metadata(Client, Topic) ->
     gen_server:cast(Client, {delete_producers_metadata, Topic}).
 
-init(St) ->
+init(#{client_id := ClientID} = St) ->
+  ok = wolff_client_sup:register_client(ClientID),
```
Comment (thalesmg) [inline src/wolff_client.erl:150]: Nit: maybe it's best to start trapping exits first and then register it?

File: src/wolff_client_sup.erl [hunk 4]
```diff
@@ -43,31 +47,57 @@ ensure_present(ClientId, Hosts, Config) ->
...
+create_clients_table() ->
+  EtsName = ?WOLFF_CLIENTS_GLOBAL_TABLE,
+  EtsOpts = [named_table, public, ordered_set, {read_concurrency, true}],
+  EtsName = ets:new(EtsName, EtsOpts),
+  ok.
+
+%% @doc Insert the client in the registration table.
+register_client(ClientId) ->
+    Pid = self(),
+    true = ets:insert(?WOLFF_CLIENTS_GLOBAL_TABLE, {ClientId, Pid}),
```
Comment (thalesmg) [inline src/wolff_client_sup.erl:97]: Nit: maybe use `ets:insert_new` to avoid overwriting a previous entry?

File: src/wolff_producers.erl [hunk 5]
```diff
@@ -18,22 +18,22 @@
...
 
--export_type([producers/0, config/0]).
+-export_type([producers/0, config/0, partitioner/0]).
 
 -include("wolff.hrl").
 -include_lib("stdlib/include/ms_transform.hrl").
 
 -opaque producers() ::
-        #{workers := #{partition() => pid()} | wolff:name(),
+        #{workers => #{partition() => pid()},
```
Comment (thalesmg) [inline src/wolff_producers.erl:36]: It seems this key was completely removed?

```suggestion
        #{
```

Comment (thalesmg) [reply src/wolff_producers.erl:36]: Actually, I see it's still returned by `start_linked_producers`.  :thinking:

File: src/wolff_producers.erl [hunk 6]
```diff
@@ -356,41 +363,39 @@ maybe_restart_producers(#{client_id := ClientId,
...
+-spec cleanup_workers_table(wolff:client_id(), wolff:topic()) -> ok.
+cleanup_workers_table(ClientId, Topic) ->
+  Ms = ets:fun2ms(fun({{C, T, _}, _}) when C =:= ClientId andalso T =:= Topic -> true end),
   ets:select_delete(?WOLFF_PRODUCERS_GLOBAL_TABLE, Ms),
   ok.
 
-find_producer_by_partition(Name, Partition) ->
-  case ets:lookup(?WOLFF_PRODUCERS_GLOBAL_TABLE, {Name, Partition}) of
-    [{_, Pid}] ->
+find_producer_by_partition(ClientId, Topic, Partition) ->
```
Comment (thalesmg) [inline src/wolff_producers.erl:372]: Just a curiosity: if `Partition = partition_count`, we get the count rather than the pid.  :joy_cat:

Comment (zmstone) [reply src/wolff_producers.erl:372]: this function is always called with a number unless when trying to find the number of partitions.

File: src/wolff_client_sup.erl (hunk 4 above)
Comment (zmstone) [reply src/wolff_client_sup.erl:97]: it should always overwrite.
client process may restart (by supervisor) after crash

File: src/wolff_producers.erl (hunk 6 above)
Comment (zmstone) [reply src/wolff_producers.erl]: add `is_integer(Partition)` guard.

**Review by qzhuyan — APPROVED** [review]

**Pull request merged**